### PR TITLE
migrate sensor support to yaq exclusively

### PR DESCRIPTION
**Pull request by untzag**
This has been a LONG time coming...

### CODE REVIEW

**Comment by ksunden** [comment]
gotta love a commit which is +0 -1721

File: pycmds/_record.py [hunk 1]
```diff
@@ -779,17 +514,17 @@ def queue_control_update(self):
...
         self.set_freerun(False)
         self.wait_until_done()
-        for device in self.devices:
-            device.close()
+        for sensor in self.sensors:
+            sensor.close()
 
     def update_cols(self, aqn):
-        for cols_type in ["data", "shots"]:
+        for cols_type in ["data"]:
```
Comment (ksunden) [inline pycmds/_record.py:527]: I _think_ this loop can just be removed entirely (i.e. remove this line and dedent the contents)

**Comment by ksunden** [comment]
The answer "ignore them for now" is acceptable (i think, though also think it may be easier to fix now, while we are thinking about it), but the modules zero_tune, abstract_tuning, and motortune all import devices still

**Comment by ksunden** [comment]
Please include the necessary changes to the default config file.

**Comment by ksunden** [comment]
While it doesn't _really_ matter, note that 39876 is the port we use on both tables (and in the config file, though disabled right now) for our spectrometers.

**Comment by ksunden** [comment]
Functionality wise it looks pretty good..

I am getting exceptions printed in pyqtgraph due to x and y not being the same shape... may have something to do with thread safety, not that big of a deal honestly, but a bit annoying.

Added a couple small commits, to remove some unused imports, then to provide sane fall back if the channels change for the scan module (if the channel you would request is not available, use the 0 index channel, as if that was not saved in state). That prevented my pycmds from starting initially, and rather than going to clear out the state, I made it so it didn't crash


I will note, while freerunning, pycmds is a bit of a cpu hog... setting a ms_wait had no effect, either as a setting or in the scan

File: pycmds/_record.py [hunk 2]
```diff
@@ -266,55 +215,21 @@ def create_data(self, inputs):
...
-                f["array"].set_fill_value = np.nan
-                f.close()
 
     def write_data(self, inputs):
-        data_arr, shots_arr = inputs
+        data_arr = inputs[0]
         # pixels --------------------------------------------------------------
         data_file = open(data_path.read(), "ab")
-        if len(data_arr.shape) == 2:  # case of multidimensional devices
+        if len(data_arr.shape) == 2:  # case of multidimensional sensors
```
Comment (ksunden) [inline pycmds/_record.py:225]: not new... but `data_arr.ndim`, and should we be handling dimensions > 2?

File: pycmds/_record.py [hunk 3]
```diff
@@ -352,160 +267,42 @@ def q(method, inputs=[]):
...
-        self.update_ui.emit()
-
-    def shutdown(self):
-        pass
-
-
-# --- gui -----------------------------------------------------------------------------------------
-
-
 class Widget(QtWidgets.QWidget):
```
Comment (ksunden) [inline pycmds/_record.py:270]: There were _two_ Widget classes previously, this looks like it is largely the _second_ one (the one that was _actually_ used previously), but at the position of the first, is this correct?

File: pycmds/_record.py [hunk 4]
```diff
@@ -710,24 +443,24 @@ def initialize_scan(self, aqn, scan_folder, destinations_list):
...
-            # apply settings from aqn to device
-            device.active = True
-            device.load_settings(aqn)
-            # record device axes, if applicable
-            if device.has_map:
-                for key in device.map_axes.keys():
+        # apply sensor settings from aqn
+        ms_wait.write(aqn.read("sensor settings", "ms wait"))
+        for sensor in self.sensors:
+            # if not aqn.has_section(sensor.name):
```
Comment (ksunden) [inline pycmds/_record.py:449]: is the `use` functionality intentionally omitted?

File: pycmds/_record.py [hunk 5]
```diff
@@ -1065,27 +712,27 @@ def create_main_tab(self):
...
         input_table = pw.InputTable()
         input_table.add("Display", None)
-        allowed_values = [device.name for device in self.control.devices]
-        self.device_combo = pc.Combo(allowed_values=allowed_values)
-        self.device_combo.updated.connect(self.on_update_device)
-        input_table.add("Device", self.device_combo)
+        allowed_values = [sensor.name for sensor in self.control.sensors]
+        self.sensor_combo = pc.Combo(allowed_values=allowed_values)
+        self.sensor_combo.updated.connect(self.on_update_sensor)
+        input_table.add("sensor", self.sensor_combo)
```
Comment (ksunden) [inline pycmds/_record.py:718]:
```suggestion
        input_table.add("Sensor", self.sensor_combo)
```

Most of our labels are capitalized

**Comment by ksunden** [comment]
Commented on a few things, but scrolled through everything and nothing seems too out of place, at least not more than it was in prior condition

File: pycmds/_record.py (hunk 2 above)
Comment (untzag) [reply pycmds/_record.py:225]: later

File: pycmds/_record.py (hunk 4 above)
Comment (untzag) [reply pycmds/_record.py:449]: not really, just taken out for now

File: pycmds/_record.py (hunk 1 above)
Comment (ksunden) [reply pycmds/_record.py:525]: @untzag ping

File: pycmds/_record.py (hunk 4 above)
Comment (ksunden) [reply pycmds/_record.py:449]: I'm confused by this statement

Comment (ksunden) [reply pycmds/_record.py:449]: ping @untzag ... wish you would have responded to this before merging...

**Pull request merged**